crates/polars-arrow/src/compute/cast/mod.rs

```diff
@@ -206,27 +211,33 @@ fn cast_list_to_fixed_size_list<O: Offset>(
             }
         }
 
-        // Build take indices for the values. This is used to fill in the null slots.
-        let mut indices =
-            MutablePrimitiveArray::<IdxSize>::with_capacity(list.values().len() + null_cnt * size);
-        for i in 0..list.len() {
-            if list.is_null(i) {
-                indices.extend_constant(size, None)
-            } else {
-                // SAFETY: we know the index is in bound.
-                let current_offset = unsafe { *offsets.get_unchecked(i) };
-                for j in 0..size {
-                    indices.push(Some(
-                        (current_offset + O::from_as_usize(j)).to_usize() as IdxSize
-                    ));
+        let list_validity = list.validity().unwrap();
+        let mut growable = make_growable(&[list.values().as_ref()], true, list.len());
+
+        if cfg!(debug_assertions) {
+            let msg = "fn cast_list_to_fixed_size_list < nullable >";
+            dbg!(msg);
+        }
+
+        for (outer_idx, x) in offsets.windows(2).enumerate() {
+            let [i, j] = x else { unreachable!() };
+            let i = i.to_usize();
+            let j = j.to_usize();
+
+            unsafe {
+                let outer_is_valid = list_validity.get_bit_unchecked(outer_idx);
+
+                if outer_is_valid {
+                    growable.extend(0, i, j - i);
+                } else {
+                    growable.extend_validity(size)
                 }
-            }
+            };
         }
-        let take_values = unsafe {
-            crate::compute::take::take_unchecked(list.values().as_ref(), &indices.freeze())
```

Comment:
Casting nullable List -> FixedSizeList, used a gather to ensure the width of the null slots